### PR TITLE
[RW-8266][risk=low] Add initial genomic file manifest publish tooling

### PR DESCRIPTION
**Pull request by calbach**
(Part 1 / x)

Add a new command `./project.rb publish-cdr-files` which generates a manifest of files to publish/copy. Initially, this tool takes a single slice through the eventual target capabilities, specifically:

- only targets CRAMs
- only generates the manifest, doesn't yet handle the copy
- eventually, I would like to upload logs / manifests to GCS for provenance

Currently outputs a CSV with the following headers:
```
source_path,ingest_path,dest_path,dest_storage_class
```

This is intended to be generic across file types, not just CRAMs.

Some of the requirements this is meeting:

- In the case of CRAMs (and some other file types), we want to rename individual files
- We need to publish the data via the ingest bucket, due to VPC-SC
- We need to specify a destination storage class

## Alternative considered

I looked a fair amount into Cloud Storage Transfer Service. Rejected for now as it seemed to lack the ability to rename files during the transfer, which is one of our requirements. It would also require testing to better understand whether it can seamlessly handle our VPC-SC ingest process, in addition to needing system-level changes.

### CODE REVIEW

File: api/db-cdr/generate-cdr/libproject/devstart.rb [hunk 1]
```diff
@@ -311,6 +313,95 @@ def publish_cdr_wgs(cmd_name, args)
...
+    "Publishing tasks to execute; defaults to all tasks: #{supported_tasks}"
+  )
+  op.opts.tier = "controlled"
+  op.add_option(
+     "--tier [tier]",
+     ->(opts, v) { opts.tier = v},
+     "The access tier associated with this CDR, e.g. controlled." +
+     "Default is controlled (WGS only exists in controlled tier, for the foreseeable future)."
+  )
+  op.add_validator ->(opts) { raise ArgumentError unless opts.project and opts.tier and opts.wgs_rids_file and opts.data_types and opts.tasks }
```
Comment (jmthibault79) [inline api/db-cdr/generate-cdr/libproject/devstart.rb:363]: We don't need to include the options which have defaults here

File: api/db-cdr/generate-cdr/libproject/devstart.rb [hunk 2]
```diff
@@ -311,6 +313,95 @@ def publish_cdr_wgs(cmd_name, args)
...
+  op.add_validator ->(opts) { raise ArgumentError.new("unsupported data types: #{opts.data_types}") unless (opts.data_types.split(",") - supported_types).empty?}
```
Comment (jmthibault79) [inline api/db-cdr/generate-cdr/libproject/devstart.rb:364]: nice arg checks

File: api/db-cdr/generate-cdr/libproject/genomic_manifests.rb [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  end
+end
+
+def _read_research_id_csvs(aw4_prefix, rids)
+  common = Common.new
+
+  ridset = rids.to_set
+
+  # See the CDR playbook for more documentation on AW4s:
+  # https://docs.google.com/document/d/1St6pG_EUFB9oRQUQaOSO7a9UPxPkQ5n4qAVyKF9j9tk/edit#
```
Comment (jmthibault79) [inline api/db-cdr/generate-cdr/libproject/genomic_manifests.rb:50]: this link did not enlighten me about AW4 - how about a brief summary comment at the top of the file?

File: api/db-cdr/generate-cdr/libproject/genomic_manifests.rb [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    else
+      aw4_str = IO.read(f)
+    end
+    if aw4_str.empty?
+      raise ArgumentError.new("failed to read content from #{f}")
+    end
+    aw4_rows = CSV.parse(aw4_str, headers: true)
+    aw4_rows.each do |row|
+      # Parse the filename for the effective datetime for downstream disambiguation.
+      row["aw4_datetime"] = filename_date
```
Comment (jmthibault79) [inline api/db-cdr/generate-cdr/libproject/genomic_manifests.rb:80]: some of these rows get filtered out just below here.  can we filter first?

File: api/db-cdr/generate-cdr/libproject/genomic_manifests.rb [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+  # Transform into a dict of research ID -> CSV::Row to dedupe
+  by_rid = {}
+  matching_aw4_rows.each do |aw4_row|
+    # Prefer the most recent row for each
+    rid = aw4_row["research_id"]
+    aw4_time = aw4_row["aw4_datetime"]
+    if by_rid.key? rid
+      common.warning "found multiple AW4 rows for #{rid}, using the most recent"
+    end
+    if not by_rid.key? rid or aw4_time < by_rid[rid]["aw4_datetime"]
```
Comment (jmthibault79) [inline api/db-cdr/generate-cdr/libproject/genomic_manifests.rb:98]: looks like this keeps the earlier datetime row, not the most recent

File: api/db-cdr/generate-cdr/libproject/genomic_manifests.rb [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+      by_rid[rid] = aw4_row
+    end
+  end
+
+  missing_rids = ridset - by_rid.keys
+  unless missing_rids.empty?
+    raise ArgumentError.new("AW4 manifests do not contain information for requested research IDs:\n" + missing_rids.join(","))
+  end
+
+  # The input resarch ID set should already account for these properties, these
```
Comment (jmthibault79) [inline api/db-cdr/generate-cdr/libproject/genomic_manifests.rb:108]: typo: research

File: api/db-cdr/generate-cdr/libproject/genomic_manifests.rb [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+  ingest_base_path = File.join(ingest_bucket, path_prefix)
+  dest_base_path = File.join(dest_bucket, path_prefix)
+
+  return wgs_aw4s.flat_map do |aw4_entry|
+    [aw4_entry["cram_path"], aw4_entry["crai_path"]].map do |source_path|
+      source_name = File.basename(source_path)
+      dest_name = source_name.sub(replace_config[:matchSourceName], replace_config[:destNameFn].call(aw4_entry["research_id"]))
+
+      if source_name == dest_name
+        raise ArgumentError.new("")
```
Comment (jmthibault79) [inline api/db-cdr/generate-cdr/libproject/genomic_manifests.rb:149]: "wrong file name format" or something?

**Review by jmthibault79 — APPROVED** [review]

File: api/db-cdr/generate-cdr/libproject/genomic_manifests.rb [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+PROD_SOURCE_CONFIG = {
+  # Optional: to speed up iteration with this script, download and run against a local directory instead.
+  :wgs_aw4_prefix => "gs://prod-drc-broad/AW4_wgs_manifest/AoU_DRCB_SEQ_"
+}
+
+
+FILE_ENVIRONMENTS = {
+  "all-of-us-rw-preprod" => {
+    :source_config => PROD_SOURCE_CONFIG,
```
[the remark this answers is not in the document]
Comment (calbach) [reply api/db-cdr/generate-cdr/libproject/genomic_manifests.rb]: PROD here is really referring to "genomics curation prod" (which doesn't have its own preprod). Tried adding a CURATION_ prefix to hopefully disambiguate.

File: api/db-cdr/generate-cdr/libproject/genomic_manifests.rb (hunk 3 above)
Comment (calbach) [reply api/db-cdr/generate-cdr/libproject/genomic_manifests.rb]: Added a longer comment to the top of the file, PTAL

File: api/db-cdr/generate-cdr/libproject/genomic_manifests.rb (hunk 5 above)
Comment (calbach) [reply api/db-cdr/generate-cdr/libproject/genomic_manifests.rb]: 🤦  - great catch

**Review by calbach — COMMENTED** [review]
Thanks!

File: api/db-cdr/generate-cdr/libproject/genomic_manifests.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (jmthibault79) [reply api/db-cdr/generate-cdr/libproject/genomic_manifests.rb]: works for me

File: api/db-cdr/generate-cdr/libproject/genomic_manifests.rb (hunk 3 above)
Comment (jmthibault79) [reply api/db-cdr/generate-cdr/libproject/genomic_manifests.rb]: very nice, thank you

File: api/db-cdr/generate-cdr/libproject/devstart.rb [hunk 9]
```diff
@@ -311,6 +313,90 @@ def publish_cdr_wgs(cmd_name, args)
...
+  )
+  op.add_option(
+    "--wgs-rids-file [file]",
+    ->(opts, v) { opts.wgs_rids_file = v},
+    "A file containing all research IDs for which WGS data should be published."
+  )
+  op.add_option(
+    "--display-version-id [version]",
+    ->(opts, v) { opts.display_version_id = v},
+    "A version 'id' suitable for display in the published GCS directory. Conventionally " +
```
Comment (jmthibault79) [inline api/db-cdr/generate-cdr/libproject/devstart.rb:333]: maybe worth also noting that this used to construct a file path?

File: api/db-cdr/generate-cdr/libproject/devstart.rb [hunk 10]
```diff
@@ -311,6 +313,90 @@ def publish_cdr_wgs(cmd_name, args)
...
+    "Publishing tasks to execute; defaults to all tasks: #{supported_tasks}"
+  )
+  op.opts.tier = "controlled"
+  op.add_option(
+     "--tier [tier]",
+     ->(opts, v) { opts.tier = v},
+     "The access tier associated with this CDR, e.g. controlled." +
+     "Default is controlled (WGS only exists in controlled tier, for the foreseeable future)."
+  )
+  op.add_validator ->(opts) { raise ArgumentError unless opts.project }
```
Comment (jmthibault79) [inline api/db-cdr/generate-cdr/libproject/devstart.rb:358]: wgs_rids_file and display_version_id are not optional

**Review by jmthibault79 — APPROVED** [review]

**Pull request merged**